### ISSUE TITLE
call does not always resolve when in map

### ISSUE THREAD

**Issue by AlexBlandin** [microsoft/pylance-release#4182]
## Environment data

-   Language Server version: Pylance 2023.3.40 (pyright 04471971)
-   OS and version: 10 21H2 19044.2728
-   Python version (& distribution if applicable, e.g. Anaconda): CPython 3.11.2 64-bit

## Code Snippet

```python
from operator import call

def foo():
  return 42

print(call(foo), *map(call, [foo]))
# 42 42
```

## Repro Steps

1. Enter code snippet
2. Observe red underline under `foo` inside `map(call, [foo])`
3. See the error claims a function without arguments cannot be `call`ed despite working perfectly in `call(foo)`.

## Expected behavior

The type of `foo` be properly handled for `call` via `map` --- indeed, it partially is, as `call` will report an error for too many positional arguments if you give `map` a second iterable, e.g. `map(call, [foo], [42])` correctly flags the error on `call` as `Function accepts too many positional parameters; expected 1 but received 2` (although the type printout is rather verbose and obscures the correct analysis), as indeed it evaluates to `TypeError: foo() takes 0 positional arguments but 1 was given`.

## Actual behavior

```py
(function) def foo() -> Literal[42]
Argument of type "list[() -> Literal[42]]" cannot be assigned to parameter "__iter1" of type "Iterable[_T1@__init__]" in function "__init__"
  Type "() -> Literal[42]" cannot be assigned to type "(**_P@call) -> Any"
Pylance(reportGeneralTypeIssues)
```

## Logs

Nothing relevant.

**Comment by erictraut** [comment]
This case is challenging for the type checker because it involves co-solving the generic types for a generic class (`map`) and a generic function (`call`) where the generic function is parameterized by a `ParamSpec`. Mypy doesn't handle this case either.

Here's a simplified version of the problem once I removed all of the complex overloads for `map.__init__`.

```python
from typing import Callable, Generic, Iterable, ParamSpec, TypeVar

_P = ParamSpec("_P")
_R = TypeVar("_R")
_S = TypeVar("_S")
_T1 = TypeVar("_T1")

class map(Generic[_S]):
    def __init__(self, func: Callable[[_T1], _S], iter1: Iterable[_T1]):
        pass

def call(obj: Callable[_P, _R], *args: _P.args, **kwargs: _P.kwargs) -> _R:
    return obj(*args, **kwargs)

def foo() -> int:
    return 42

# Mypy: Argument 1 to "map" has incompatible type "Callable[[Callable[_P, _R], **_P], _R]"; expected "Callable[[Callable[[], int]], _R]"
# Pyright: Argument of type "list[() -> int]" cannot be assigned to parameter "__iter1" of type "Iterable[_T1@__init__]" in function "__init__"
result = map(call, [foo])

reveal_type(result)
```

With my fix in place, pyright no longer generates the error, and it produces a type result of `map[int]` as expected.

This will be included in an upcoming release of pylance.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2023.4.11, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023411-5-April-2023-PreRelease)
